### DISCUSSION TITLE
netbox custom script - csv file permission denied

### DISCUSSION THREAD

**Discussion by nikunj-tele** [netbox-community/netbox#15526] [Help Wanted!]
Hi All,
I have written the script to generate the inventory output but when I tried to download this output in csv file it gives the following error , my script is working fine if it generate the output but only csv failed

An exception occurred: PermissionError: [Errno 13] Permission denied: 'netbox_custom_inventory.csv'

Traceback (most recent call last):
  File "/opt/netbox/netbox/extras/scripts.py", line 504, in _run_script
    script.output = script.run(data=data, commit=commit)
  File "/opt/netbox/netbox/scripts/netbox-invscript.py", **line 32, in run**
    with open(OUTPUT_FILE, "w", newline="") as csvfile:
**PermissionError: [Errno 13] Permission denied: 'netbox_custom_inventory.csv'**

snippest of the code is as below 

```
def run(self, data, commit):
	
		OUTPUT_FILE = 'netbox_custom_inventory.csv'  ## 
		fields = [
				"Name",
				"Serial,Site",
				
					........,"location",]
				with open(OUTPUT_FILE, "w", newline="") as csvfile:  ## <-- this is line 32 
			writer = csv.DictWriter(csvfile, fieldnames=fields)
			writer.writeheader()
						for device_name in devices:
			try:
				device = Device.objects.get(name=device_name)  # Get device by name
								writer.writerow([device.name,str(device.serial),str(device.site),str(device.custom_field_data["Responsible_Squad"]),.....so on)])
				print (f"CSV report saved to {OUTPUT_FILE}")
			except:
				return f"Error: Device with name '{device_name}' not found."

```
please ignore wrong indentation as coping code messed  it.!

Unfortunately , I donot have access on the production server to create and change the ownership/permission of the csv file.
Is there any way to write the output into the csv file and download it ?
is is possible to generate csv file using custom script ?

Thank you in advance.

**Comment by candlerb** [comment]
As the error quite clearly says, this is a permissions problem. Your custom script can only write to directories which are owned by the "netbox" user  (i.e. it's not a problem with the file, it's the directory you're trying to write to)

But in this case there's no need to write to any file on the server.  Just get your custom script to return the CSV as a string, by using StringIO, and the user will get it in their browser. There's an example custom script [here](https://github.com/netbox-community/customizations/blob/master/scripts/power_summary.py) which you can copy.

**Reply by nikunj-tele** [reply]
Hi @candlerb ,
Thank  you for your response, I already tried sciprt using StringIO but it give the csv output on screen so how can I export the csv output to excel? 
is it possible?

**Reply by candlerb** [reply]
Copy-paste from the screen into a file is the simple answer. I don't think scripts can set "content-disposition" which would allow the output to be downloadable.
